server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -582,50 +596,97 @@ static SearchResponseMerger createSearchResponseMerger(
         final AtomicReference<Exception> exceptions = new AtomicReference<>();
         for (Map.Entry<String, OriginalIndices> entry : remoteIndicesByCluster.entrySet()) {
             final String clusterAlias = entry.getKey();
             boolean skipUnavailable = remoteClusterService.isSkipUnavailable(clusterAlias);
-            Client clusterClient = remoteClusterService.getRemoteClusterClient(threadPool, clusterAlias);
-            final String[] indices = entry.getValue().indices();
-            ClusterSearchShardsRequest searchShardsRequest = new ClusterSearchShardsRequest(indices).indicesOptions(indicesOptions)
-                .local(true)
-                .preference(preference)
-                .routing(routing);
-            clusterClient.admin()
-                .cluster()
-                .searchShards(
-                    searchShardsRequest,
-                    new CCSActionListener<ClusterSearchShardsResponse, Map<String, ClusterSearchShardsResponse>>(
-                        clusterAlias,
-                        skipUnavailable,
-                        responsesCountDown,
-                        skippedClusters,
-                        exceptions,
-                        listener
-                    ) {
-                        @Override
-                        void innerOnResponse(ClusterSearchShardsResponse clusterSearchShardsResponse) {
-                            searchShardsResponses.put(clusterAlias, clusterSearchShardsResponse);
-                        }
+            TransportSearchAction.CCSActionListener<SearchShardsResponse, Map<String, SearchShardsResponse>> singleListener =
+                new TransportSearchAction.CCSActionListener<>(
+                    clusterAlias,
+                    skipUnavailable,
+                    responsesCountDown,
+                    skippedClusters,
+                    exceptions,
+                    listener
+                ) {
+                    @Override
+                    void innerOnResponse(SearchShardsResponse searchShardsResponse) {
+                        searchShardsResponses.put(clusterAlias, searchShardsResponse);
+                    }
 
-                        @Override
-                        Map<String, ClusterSearchShardsResponse> createFinalResponse() {
-                            return searchShardsResponses;
-                        }
+                    @Override
+                    Map<String, SearchShardsResponse> createFinalResponse() {
+                        return searchShardsResponses;
                     }
-                );
+                };
+            final StepListener<Void> ensureConnectedListener = new StepListener<>();
+            if (skipUnavailable) {
+                ensureConnectedListener.onResponse(null);
+            } else {
+                remoteClusterService.ensureConnected(clusterAlias, ensureConnectedListener);
+            }
```

Comment:
I don't understand what is happening in this if/else block. Reading down into the `remoteClusterService .ensureConnected` code, it looks like the service will check if it is connected to the remote cluster and if so, it calls the onResponse callback of the listener with `null` as you do on line 639. Importantly, it calls listener.onFailure(Exception) if the connection fails.

I get the calling of `listener.onFailure` only if `skipUnavailable==false`, but I don't understand the point of calling `onResponse(null)` on this listener. Especially on line 638 when you know that all you have is an empty StepListener that (I don't think?) is configured to do anything at that point in time.

Is this `onResponse(null)` call necessary? Or is just there for future proofing? I wonder if it needs some code comments to explain why it is being done.

Reply:
Calling `RemoteClusterService#ensureConnected` can take time as we need to re-establish the connection to the remote cluster if we don't have one. We prefer to fail fast search requests with `skip_unavailable=true` in this case. The detail can be found in https://github.com/elastic/elasticsearch/pull/80589 and https://github.com/elastic/elasticsearch/issues/74773. Happy to walk through if you need more context.

> I don't understand the point of calling `onResponse(null)` on this listener.

`onResponse(null)` means we are skipping `RemoteClusterService#ensureConnected` when `skip_unavailble=true`

Reply:
The logic here is similar to what we have in RemoteClusterAwareClient.

Reply:
OK, but if you want to short-circuit should you put a return statement in the `if` block after line 639?